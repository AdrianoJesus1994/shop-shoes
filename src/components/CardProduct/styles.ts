import {Dimensions} from 'react-native';
import theme from './../../utils/theme';
import styled from 'styled-components/native';

export const Container = styled.View`
  margin: 16px 8px;
  flex: 1;
`;

export const ImageContent = styled.View`
  border-radius: 10px;
  overflow: hidden;
`;

export const ContentStars = styled.View`
  flex: 1;
  flex-direction: row;
`;

export const ProductPrice = styled.Text`
  font-family: 'Montserrat-Bold';
  font-size: 22px;
  color: #999;
`;

export const ProductName = styled.Text`
  font-size: 18px;
  font-family: 'Montserrat-Bold';
`;

export const ProductBody = styled.View`
  flex: 1;
  text-align: center;
  margin: 8px;
`;
