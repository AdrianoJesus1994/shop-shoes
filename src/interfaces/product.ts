export interface Product {
  principal_image: string;
  other_images: Array<string>;
  product_name: string;
  product_description: string;
  price: string | number;
  sale_price?: string;
  evaluations: Array<Evaluation>;
  comments: Array<Comment>;
}

export interface Comment {
  author_name: string;
  comment_text: string;
}

export interface Evaluation {
  rate: RateValue;
}

export enum RateValue {
  one = 1,
  two = 2,
  tree = 3,
  four = 4,
  five = 5,
}
