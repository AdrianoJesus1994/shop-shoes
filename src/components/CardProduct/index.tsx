import React from 'react';
import {View, Image} from 'react-native';
import {Product} from '../../interfaces/product';
import MDIcon from 'react-native-vector-icons/MaterialCommunityIcons';

import {
  Container,
  ImageContent,
  ContentStars,
  ProductBody,
  ProductName,
  ProductPrice,
} from './styles';
import theme from '../../utils/theme';
import {useNavigation} from '@react-navigation/native';
import route_names from '../../utils/route_names';

interface CardProductProps {
  product: Product;
  onTap?: (product_id: number) => void;
}

const CardProduct: React.FC<CardProductProps> = ({product, onTap}) => {
  function handleOpenPageDetail() {
    if (onTap) onTap(Number(product.id));
    //navigation.navigate(route_names.DETAILS, {product_id: product.id});
  }

  return (
    <Container onPress={handleOpenPageDetail}>
      <ImageContent>
        <Image
          source={{uri: product.principal_image}}
          style={{aspectRatio: 1 / 1}}
        />
      </ImageContent>
      <ProductBody>
        <ContentStars>
          <MDIcon name="star" size={22} color={theme.darkOrange} />
          <MDIcon name="star" size={22} color={theme.darkOrange} />
          <MDIcon name="star" size={22} color={theme.darkOrange} />
          <MDIcon name="star" size={22} color={theme.darkOrange} />
          <MDIcon name="star" size={22} color={theme.light} />
        </ContentStars>
        <ProductName>{product.product_name}</ProductName>
        <ProductPrice>R$ {product.price.toFixed(2)}</ProductPrice>
      </ProductBody>
    </Container>
  );
};

export default CardProduct;
