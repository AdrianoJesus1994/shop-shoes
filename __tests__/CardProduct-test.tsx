import React from 'react';
import {TouchableOpacity} from 'react-native';

import CardProduct from './../src/components/CardProduct';
import renderer from 'react-test-renderer';
import {Product} from '../src/interfaces/product';

describe('Teste CardProduct component', () => {
  it('should present product when params passed corretly', () => {
    const product: Product = {
      id: 5,
      product_name: 'Tenis Nike',
      product_description: 'Tenis de alta qualidade para os melhores clientes.',
      price: 150.5,
      principal_image:
        'https://static.netshoes.com.br/produtos/tenis-nike-revolution-5-masculino/36/HZM-1731-036/HZM-1731-036_zoom1.jpg?ts=1569489111?ims=544x',
      other_images: [
        'https://static.netshoes.com.br/produtos/tenis-de-caminhada-leve-confortavel/06/E74-0492-006/E74-0492-006_zoom1.jpg?ts=1586961574&ims=544x',
        'https://static.netshoes.com.br/produtos/tenis-de-caminhada-leve-confortavel/06/E74-0492-006/E74-0492-006_zoom1.jpg?ts=1586961574&ims=544x',
      ],
      evaluations: [],
      comments: [],
    };

    const tree = renderer.create(<CardProduct product={product} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call function in prop to action when pressed', () => {
    const mockFn = jest.fn();

    const product: Product = {
      id: 5,
      product_name: 'Tenis Nike',
      product_description: 'Tenis de alta qualidade para os melhores clientes.',
      price: 150.5,
      principal_image:
        'https://static.netshoes.com.br/produtos/tenis-nike-revolution-5-masculino/36/HZM-1731-036/HZM-1731-036_zoom1.jpg?ts=1569489111?ims=544x',
      other_images: [
        'https://static.netshoes.com.br/produtos/tenis-de-caminhada-leve-confortavel/06/E74-0492-006/E74-0492-006_zoom1.jpg?ts=1586961574&ims=544x',
        'https://static.netshoes.com.br/produtos/tenis-de-caminhada-leve-confortavel/06/E74-0492-006/E74-0492-006_zoom1.jpg?ts=1586961574&ims=544x',
      ],
      evaluations: [],
      comments: [],
    };

    const testRenderer = renderer.create(
      <CardProduct product={product} onTap={mockFn} />,
    );
    const cardTest = testRenderer.root;

    const button = cardTest.findByType(TouchableOpacity);

    renderer.act(button.props.onPress);

    expect(mockFn).toBeCalled();
    expect(mockFn).toHaveBeenCalledTimes(1);
  });
});
