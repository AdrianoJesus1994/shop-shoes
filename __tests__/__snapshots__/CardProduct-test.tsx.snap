// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Teste CardProduct component should present product when params passed corretly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "marginBottom": 16,
      "marginLeft": 8,
      "marginRight": 8,
      "marginTop": 16,
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 10,
          "overflow": "hidden",
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": "https://static.netshoes.com.br/produtos/tenis-nike-revolution-5-masculino/36/HZM-1731-036/HZM-1731-036_zoom1.jpg?ts=1569489111?ims=544x",
        }
      }
      style={
        Object {
          "aspectRatio": 1,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginBottom": 8,
          "marginLeft": 8,
          "marginRight": 8,
          "marginTop": 8,
          "textAlign": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#ff8c00",
              "fontSize": 22,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰓎
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#ff8c00",
              "fontSize": 22,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰓎
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#ff8c00",
              "fontSize": 22,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰓎
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#ff8c00",
              "fontSize": 22,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰓎
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#cdcdcd",
              "fontSize": 22,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰓎
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Montserrat-Bold",
            "fontSize": 18,
          },
        ]
      }
    >
      Tenis Nike
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#999",
            "fontFamily": "Montserrat-Bold",
            "fontSize": 22,
          },
        ]
      }
    >
      R$ 
      150.50
    </Text>
  </View>
</View>
`;
